hypershift-operator/controllers/machinedeployment/controller.go

```diff
@@ -0,0 +1,279 @@
+
+	// Label used to detect if CPO natively supports scale-from-zero
+	controlPlaneOperatorSupportsAWSAutoscalingFromZeroLabel = "io.openshift.hypershift.control-plane-operator-supports-aws-autoscaling-from-zero"
+)
+
+// Reconciler reconciles MachineDeployments for scale-from-zero autoscaling.
+type Reconciler struct {
+	Client             client.Client
+	Log                logr.Logger
+	CredentialsFile    string
+	RegionCache        awsclient.RegionCache
+	InstanceTypesCache InstanceTypesCache
+
+	// For detecting CPO capabilities to skip workaround when CPO supports it natively
+	ImageMetadataProvider   supportutil.ImageMetadataProvider
+	ReleaseProvider         releaseinfo.Provider
+	HypershiftOperatorImage string
+
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+// SetupWithManager creates a new controller for a manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineDeployment{}).
+		Build(r)
+
+	if err != nil {
+		return fmt.Errorf("failed setting up with a controller manager: %w", err)
+	}
+
+	r.recorder = mgr.GetEventRecorderFor("machinedeployment-controller")
+	r.scheme = mgr.GetScheme()
+	return nil
+}
+
+// Reconcile implements controller runtime Reconciler interface.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("machinedeployment", req.Name, "namespace", req.Namespace)
+	logger.V(3).Info("Reconciling")
+
+	machineDeployment := &clusterv1.MachineDeployment{}
+	if err := r.Client.Get(ctx, req.NamespacedName, machineDeployment); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Ignore deleted MachineDeployments, this can happen when foregroundDeletion
+	// is enabled
+	if !machineDeployment.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	originalMachineDeploymentToPatch := client.MergeFrom(machineDeployment.DeepCopy())
+
+	result, err := r.reconcile(ctx, machineDeployment)
+	if err != nil {
+		logger.Error(err, "Failed to reconcile MachineDeployment")
+		r.recorder.Eventf(machineDeployment, corev1.EventTypeWarning, "ReconcileError", "%v", err)
+		// we don't return here so we want to attempt to patch the machine regardless of an error.
+	}
+
+	if err := r.Client.Patch(ctx, machineDeployment, originalMachineDeploymentToPatch); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to patch machineDeployment: %v", err)
+	}
+
+	return result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, machineDeployment *clusterv1.MachineDeployment) (ctrl.Result, error) {
+	klog.V(3).Infof("%v: Reconciling MachineDeployment", machineDeployment.Name)
```

Comment:
please use the controller-runtime logger instead of klog
you can fetch the logger from the context with ctrl.LoggerFrom(ctx)